File: ToC_2026/8/question4/Haskell/2.hs
//数値の等価演算子 eq
iszero :: Int -> Bool
iszero m = m == 0

pi_ :: Int -> Int
pi_ m = if m == 0 then 0 else m - 1

eq :: Int -> Int -> Bool
eq m n = if iszero m
         then iszero n
         else if iszero n
              then False
              else eq (pi_ m) (pi_ n)

main :: IO ()
main = do
    putStrLn "Enter two numbers: "
    [m, n] <- map read . words <$> getLine
    if eq m n
        then putStrLn $ show m ++ " and " ++ show n ++ " are equal."
        else putStrLn $ show m ++ " and " ++ show n ++ " are not equal."

File: ToC_2026/8/question4/C/2.c
//数値の等価演算子 eq
#include <stdio.h>
int iszero(int m) { return m == 0; }
int pi(int m) { if (m == 0) return 0; return m - 1; }

int eq(int m, int n) {
    if (iszero(m)) {
        return iszero(n);
    } else if (iszero(n)) {
        return 0; 
    } else {
        return eq(pi(m), pi(n));
    }
}

int main() {
    int m, n;
    printf("Enter two numbers: ");
    scanf("%d %d", &m, &n);
    if (eq(m, n)) {
        printf("%d and %d are equal.\n", m, n);
    } else {
        printf("%d and %d are not equal.\n", m, n);
    }
    return 0;
}